File: app/src/main/java/com/simplemobiletools/flashlight/activities/WidgetConfigureActivity.java
package com.simplemobiletools.flashlight.activities;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ImageView;
import android.widget.RemoteViews;
import android.widget.SeekBar;

import com.simplemobiletools.flashlight.Constants;
import com.simplemobiletools.flashlight.MyWidgetProvider;
import com.simplemobiletools.flashlight.R;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import yuku.ambilwarna.AmbilWarnaDialog;

public class WidgetConfigureActivity extends AppCompatActivity {
    @BindView(R.id.config_widget_seekbar) SeekBar mWidgetSeekBar;
    @BindView(R.id.config_widget_color) View mWidgetColorPicker;
    @BindView(R.id.config_image) ImageView mImage;

    private static float mWidgetAlpha;
    private static int mWidgetId;
    private static int mWidgetColor;
    private static int mWidgetColorWithoutTransparency;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setResult(RESULT_CANCELED);
        setContentView(R.layout.widget_config);
        ButterKnife.bind(this);
        initVariables();

        final Intent intent = getIntent();
        final Bundle extras = intent.getExtras();
        if (extras != null)
            mWidgetId = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);

        if (mWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID)
            finish();
    }

    private void initVariables() {
        final SharedPreferences prefs = getSharedPreferences(Constants.PREFS_KEY, Context.MODE_PRIVATE);
        mWidgetColor = prefs.getInt(Constants.WIDGET_COLOR, 1);
        if (mWidgetColor == 1) {
            mWidgetColor = getResources().getColor(R.color.colorPrimary);
            mWidgetAlpha = 1.f;
        } else {
            mWidgetAlpha = Color.alpha(mWidgetColor) / (float) 255;
        }

        mWidgetColorWithoutTransparency = Color.rgb(Color.red(mWidgetColor), Color.green(mWidgetColor), Color.blue(mWidgetColor));
        mWidgetSeekBar.setOnSeekBarChangeListener(seekbarChangeListener);
        mWidgetSeekBar.setProgress((int) (mWidgetAlpha * 100));
        updateColors();
    }

    @OnClick(R.id.config_save)
    public void saveConfig() {
        final AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(this);
        final RemoteViews views = new RemoteViews(getPackageName(), R.layout.widget);
        appWidgetManager.updateAppWidget(mWidgetId, views);

        storeWidgetColors();
        requestWidgetUpdate();

        final Intent resultValue = new Intent();
        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mWidgetId);
        setResult(RESULT_OK, resultValue);
        finish();
    }

    @OnClick(R.id.config_widget_color)
    public void pickBackgroundColor() {
        AmbilWarnaDialog dialog = new AmbilWarnaDialog(this, mWidgetColorWithoutTransparency, new AmbilWarnaDialog.OnAmbilWarnaListener() {
            @Override
            public void onCancel(AmbilWarnaDialog dialog) {
            }

            @Override
            public void onOk(AmbilWarnaDialog dialog, int color) {
                mWidgetColorWithoutTransparency = color;
                updateColors();
            }
        });

        dialog.show();
    }

    private void storeWidgetColors() {
        final SharedPreferences prefs = getSharedPreferences(Constants.PREFS_KEY, Context.MODE_PRIVATE);
        prefs.edit().putInt(Constants.WIDGET_COLOR, mWidgetColor).apply();
    }

    private void requestWidgetUpdate() {
        final Intent intent = new Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE, null, this, MyWidgetProvider.class);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, new int[]{mWidgetId});
        sendBroadcast(intent);
    }

    private void updateColors() {
        mWidgetColor = adjustAlpha(mWidgetColorWithoutTransparency, mWidgetAlpha);
        mWidgetColorPicker.setBackgroundColor(mWidgetColor);
        mImage.getDrawable().mutate().setColorFilter(mWidgetColor, PorterDuff.Mode.SRC_IN);
    }

    private SeekBar.OnSeekBarChangeListener seekbarChangeListener = new SeekBar.OnSeekBarChangeListener() {
        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            mWidgetAlpha = (float) progress / (float) 100;
            updateColors();
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {

        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {

        }
    };

    private int adjustAlpha(int color, float factor) {
        final int alpha = Math.round(Color.alpha(color) * factor);
        final int red = Color.red(color);
        final int green = Color.green(color);
        final int blue = Color.blue(color);
        return Color.argb(alpha, red, green, blue);
    }
}

File: app/src/main/kotlin/com/simplemobiletools/flashlight/activities/WidgetConfigureActivity.kt
package com.simplemobiletools.flashlight.activities

import android.app.Activity
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.ImageView
import android.widget.RemoteViews
import android.widget.SeekBar
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.simplemobiletools.commons.helpers.PREFS_KEY
import com.simplemobiletools.flashlight.R
import com.simplemobiletools.flashlight.helpers.MyWidgetProvider
import com.simplemobiletools.flashlight.helpers.WIDGET_COLOR
import yuku.ambilwarna.AmbilWarnaDialog

class WidgetConfigureActivity : AppCompatActivity() {
    @BindView(R.id.config_widget_seekbar) internal var mWidgetSeekBar: SeekBar? = null
    @BindView(R.id.config_widget_color) internal var mWidgetColorPicker: View? = null
    @BindView(R.id.config_image) internal var mImage: ImageView? = null

    private val seekbarChangeListener = object : SeekBar.OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            mWidgetAlpha = progress.toFloat() / 100.toFloat()
            updateColors()
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {

        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {

        }
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setResult(Activity.RESULT_CANCELED)
        setContentView(R.layout.widget_config)
        ButterKnife.bind(this)
        initVariables()

        val intent = intent
        val extras = intent.extras
        if (extras != null)
            mWidgetId = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)

        if (mWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID)
            finish()
    }

    private fun initVariables() {
        val prefs = getSharedPreferences(PREFS_KEY, Context.MODE_PRIVATE)
        mWidgetColor = prefs.getInt(WIDGET_COLOR, 1)
        if (mWidgetColor == 1) {
            mWidgetColor = resources.getColor(R.color.colorPrimary)
            mWidgetAlpha = 1f
        } else {
            mWidgetAlpha = Color.alpha(mWidgetColor) / 255.toFloat()
        }

        mWidgetColorWithoutTransparency = Color.rgb(Color.red(mWidgetColor), Color.green(mWidgetColor), Color.blue(mWidgetColor))
        mWidgetSeekBar!!.setOnSeekBarChangeListener(seekbarChangeListener)
        mWidgetSeekBar!!.progress = (mWidgetAlpha * 100).toInt()
        updateColors()
    }

    @OnClick(R.id.config_save)
    fun saveConfig() {
        val appWidgetManager = AppWidgetManager.getInstance(this)
        val views = RemoteViews(packageName, R.layout.widget)
        appWidgetManager.updateAppWidget(mWidgetId, views)

        storeWidgetColors()
        requestWidgetUpdate()

        val resultValue = Intent()
        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mWidgetId)
        setResult(Activity.RESULT_OK, resultValue)
        finish()
    }

    @OnClick(R.id.config_widget_color)
    fun pickBackgroundColor() {
        val dialog = AmbilWarnaDialog(this, mWidgetColorWithoutTransparency, object : AmbilWarnaDialog.OnAmbilWarnaListener {
            override fun onCancel(dialog: AmbilWarnaDialog) {}

            override fun onOk(dialog: AmbilWarnaDialog, color: Int) {
                mWidgetColorWithoutTransparency = color
                updateColors()
            }
        })

        dialog.show()
    }

    private fun storeWidgetColors() {
        val prefs = getSharedPreferences(PREFS_KEY, Context.MODE_PRIVATE)
        prefs.edit().putInt(WIDGET_COLOR, mWidgetColor).apply()
    }

    private fun requestWidgetUpdate() {
        val intent = Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE, null, this, MyWidgetProvider::class.java)
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, intArrayOf(mWidgetId))
        sendBroadcast(intent)
    }

    private fun updateColors() {
        mWidgetColor = adjustAlpha(mWidgetColorWithoutTransparency, mWidgetAlpha)
        mWidgetColorPicker!!.setBackgroundColor(mWidgetColor)
        mImage!!.background.mutate().setColorFilter(mWidgetColor, PorterDuff.Mode.SRC_IN)
    }

    private fun adjustAlpha(color: Int, factor: Float): Int {
        val alpha = Math.round(Color.alpha(color) * factor)
        val red = Color.red(color)
        val green = Color.green(color)
        val blue = Color.blue(color)
        return Color.argb(alpha, red, green, blue)
    }

    companion object {

        private var mWidgetAlpha: Float = 0.toFloat()
        private var mWidgetId: Int = 0
        private var mWidgetColor: Int = 0
        private var mWidgetColorWithoutTransparency: Int = 0
    }
}
